### PR TITLE
Unify share entry and flow

### PR DESCRIPTION
**Pull request by ryanlntn**
I started this as an experiment to see if we could just use `index.js` as our entry point. After some patching of the native share extension code it does work and provides us with the follow advantages:

1. We now only have one source of setup for the app and share extension. Gone are the days of that breaking out from under us.
2. We now have boot splash and firebase within the share extension (note that FIRTrace doesn't support app extensions so we don't have startup time metrics here instead I just stubbed it out)
3. Because we now have firebase we also have ErrorBoundary working in the share extension.
4. We now pass share extension data via route params just like the Android flow making the initial share handling the only platform specific code.
5. Building on 4 we don't need a new navigator and separate navigation logic for Android/iOS anymore.
6. This also gives us more options for handling offline behavior since authenticating within the share extension should be possible now.
7. This also updates text share handling to drop the shared text directly in the message field.

Trello:
https://trello.com/c/gqu6Q801/156-share-extension-and-app-initialization-code-sharing
https://trello.com/c/a0MCWIiw/155-determine-and-implement-text-sharing-behavior

### CODE REVIEW

File: patches/react-native-share-menu+6.0.0.patch [hunk 1]
```diff
@@ -1,13 +1,94 @@
...
+index f42bce6..ab00813 100644
 --- a/node_modules/react-native-share-menu/ios/ReactShareViewController.swift
 +++ b/node_modules/react-native-share-menu/ios/ReactShareViewController.swift
 @@ -13,7 +13,7 @@ class ReactShareViewController: ShareViewController, RCTBridgeDelegate, ReactSha
    func sourceURL(for bridge: RCTBridge!) -> URL! {
  #if DEBUG
      return RCTBundleURLProvider.sharedSettings()?
 -      .jsBundleURL(forBundleRoot: "index.share", fallbackResource: nil)
-+      .jsBundleURL(forBundleRoot: "index.share")
++      .jsBundleURL(forBundleRoot: "index")
```
Comment (lindboe) [inline patches/react-native-share-menu+6.0.0.patch:46]: Is there a reason anyone might ever need to use index.share.js?

File: patches/react-native-share-menu+6.0.0.patch [hunk 2]
```diff
@@ -1,13 +1,94 @@
...
+-      }
+-
+-      let red = backgroundColorConfig[COLOR_RED_KEY] as? Float ?? 1
+-      let green = backgroundColorConfig[COLOR_GREEN_KEY] as? Float ?? 1
+-      let blue = backgroundColorConfig[COLOR_BLUE_KEY] as? Float ?? 1
+-      let alpha = backgroundColorConfig[COLOR_ALPHA_KEY] as? Float ?? 1
+-
+-      rootView.backgroundColor = UIColor(red: CGFloat(red), green: CGFloat(green), blue: CGFloat(blue), alpha: CGFloat(alpha))
+-    }
++    RCTBootSplash.initWithStoryboard("MainInterface", rootView: rootView)
```
Comment (lindboe) [inline patches/react-native-share-menu+6.0.0.patch:68]: Ah, is this why the changes to the storyboard file? Should that be part of the patch so everyone using the library would get this? I honestly don't know anything about storyboard. What's the motivation for this change?

File: patches/react-native-share-menu+6.0.0.patch [hunk 3]
```diff
@@ -1,13 +1,94 @@
...
+ import RNShareMenu
++import Firebase
+ 
+ class ShareViewController: SLComposeServiceViewController {
+   var hostAppId: String?
+@@ -22,6 +23,10 @@ class ShareViewController: SLComposeServiceViewController {
+   override func viewDidLoad() {
+     super.viewDidLoad()
+     
++    if FirebaseApp.app() == nil {
```
Comment (lindboe) [inline patches/react-native-share-menu+6.0.0.patch:88]: Hmm, this is another one that seems difficult to upstream. That said, it does seem useful to have the error boundary working... but at the same time, the error boundary component itself was designed to make sense within the main app (where you'd reload it).

If we're going to include the error boundary's functionality in terms of crashlytics reporting, I think it makes sense to work that into the design and talk about how it should look in the share extension.

If we did want to go down this road, in terms of the library changes, I imagine what we'd need to do is create classes meant to be extended by custom classes in the project, instead of just directly adding the file? Like... most libraries.

File: src/libs/Share/navigateToShare.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import isArray from 'lodash/isArray';
+
+import ROUTES from '../../ROUTES';
+import Navigation from '../Navigation/Navigation';
+
+const navigateToShare = (share) => {
+    if (!share || !share.data) return;
+    if (isArray(share.data) && share.data.length === 0) return;
```
Comment (lindboe) [inline src/libs/Share/navigateToShare.js:8]: note, for the final code it might be nice to have a `shareDataExists` utility fn

File: src/pages/ShareMessagePage.js [hunk 5]
```diff
@@ -9,27 +9,27 @@ import ScreenWrapper from '../components/ScreenWrapper';
...
 import * as Report from '../libs/actions/Report';
 import styles from '../styles/styles';
 
 function ShareMessagePage(props) {
     const reportID = props.route.params.reportID;
-    const [attachment, setAttachment] = useState(props.route.params.share);
-    const [message, setMessage] = useState('');
+    const attachment = props.route.params.share;
+    const source = attachment.data;
+    const uri = attachment.data;
```
Comment (lindboe) [inline src/pages/ShareMessagePage.js:22]: Looks like a duplicate of `source`?

File: src/pages/ShareMessagePage.js [hunk 6]
```diff
@@ -9,27 +9,27 @@ import ScreenWrapper from '../components/ScreenWrapper';
...
+    const name = source.split('/').pop();
```
Comment (lindboe) [inline src/pages/ShareMessagePage.js:23]: note, also for polish, would be good to have utility functions in lib for share data

File: src/pages/ShareMessagePage.js [hunk 7]
```diff
@@ -9,27 +9,27 @@ import ScreenWrapper from '../components/ScreenWrapper';
...
-        } else {
-            setAttachment(props.route.params.share);
-        }
-    }, [props.route.params.share]);
+    const [message, setMessage] = useState(isTextShare ? source : '');
 
-    // TODO: fix these
-    const close = Platform.select({ios: () => ShareMenuReactView.dismissExtension(), default: () => Navigation.dismissModal()});
-    const goBack = Platform.select({ios: () => console.log('TODO: fix goBack'), default: () => Navigation.goBack()});
+    const dismiss = Platform.select({
```
Comment (lindboe) [inline src/pages/ShareMessagePage.js:29]: should libify this with the platform extensions so we can avoid the Platform.select

**Comment by lindboe** [comment]
FYI, I saw another issue after testing just now: I wasn't signed in, and saw the sign-in page, and it was very apparent the correct fonts weren't loaded. So I think we might need another task to bring in the font assets.

Share extension:
Screenshot 2023-07-27 at 8 21 51 PM

Main app:
Screenshot 2023-07-27 at 8 22 00 PM

**Comment by lindboe** [comment]
Just tested this out on iOS trying to share a photo, and it looks like it tried to share it as a text link? Not sure if that's just how this would normally work from the way I'm doing it.

On the test prior to this that I didn't record, instead of the `https` link, it used a local `file` link, I'm not sure why the difference, I thought I took the same actions.

Also of note: I think the splash screen might not be beneficial here. In the video you can see the UI render before the splash screen shows.

https://github.com/infinitered/ExpensifyApp/assets/5252268/5987caff-858c-43f1-8e10-c2978c39225f

**Review by lindboe — COMMENTED** [review]
Tested out iOS and have a few questions about what's expected at this point, so just leaving comments for now

**Comment by ryanlntn** [comment]
> Just tested this out on iOS trying to share a photo, and it looks like it tried to share it as a text link? Not sure if that's just how this would normally work from the way I'm doing it.

Just tested sharing to Messages and it works the same way when sharing from web. I think we're looking at OS/App specific behavior.

File: index.js [hunk 8]
```diff
@@ -2,11 +2,12 @@
  * @format
  */
 
-import 'react-native-gesture-handler';
 import {AppRegistry} from 'react-native';
+import 'react-native-gesture-handler';
```
[the remark this answers is not in the document]
Comment (ryanlntn) [reply index.js:6]: Format on save and a combination of prettier and organize imports.

File: ios/share/Base.lproj/MainInterface.storyboard [hunk 9]
```diff
@@ -11,16 +11,34 @@
         <scene sceneID="ceB-am-kn3">
             <objects>
                 <viewController id="j1y-V4-xli" customClass="ReactShareViewController" customModule="share" customModuleProvider="target" sceneMemberID="viewController">
-                    <view key="view" opaque="NO" contentMode="scaleToFill" id="wbc-yd-nQP">
+                    <view key="view" autoresizesSubviews="NO" userInteractionEnabled="NO" contentMode="scaleToFill" id="I1y-SQ-EVP">
                         <rect key="frame" x="0.0" y="0.0" width="393" height="852"/>
-                        <autoresizingMask key="autoresizingMask" widthSizable="YES" heightSizable="YES"/>
-                        <viewLayoutGuide key="safeArea" id="1Xd-am-t49"/>
-                        <color key="backgroundColor" red="0.0" green="0.0" blue="0.0" alpha="0.0" colorSpace="custom" customColorSpace="sRGB"/>
+                        <autoresizingMask key="autoresizingMask"/>
```
[the remark this answers is not in the document]
Comment (ryanlntn) [reply ios/share/Base.lproj/MainInterface.storyboard:16]: It's replaced with the contents of Bootsplash. I might actually try to see if I can just reuse Bootsplash instead. I tried and failed before but it should be possible.

File: patches/react-native-share-menu+6.0.0.patch (hunk 1 above)
Comment (ryanlntn) [reply patches/react-native-share-menu+6.0.0.patch:46]: Kind of comes down to code organization preferences. It makes it easier/more declarative to specifically setup the share extension. In most apps that would be preferred. In our case since we have a ludicrously tangled mess it makes more sense to just reuse index.js.

File: patches/react-native-share-menu+6.0.0.patch (hunk 2 above)
Comment (ryanlntn) [reply patches/react-native-share-menu+6.0.0.patch:68]: Mostly just trying to replicate as much of the setup from AppDelegate as we can. In release mode there's no pause waiting for metro so you just get a very brief flash of this before you're already in the app. We might be able to remove it just need to make sure the calls to hide the bootsplash later don't blow up.

File: patches/react-native-share-menu+6.0.0.patch (hunk 3 above)
Comment (ryanlntn) [reply patches/react-native-share-menu+6.0.0.patch:88]: Yeah there are a couple approaches we could take. We could instruct people to "Copy items if needed" instead of deliberately not doing that. Or maybe we could provide a way to hook into these particular lifecycle methods via swift extension or something similar. Either way I think providing a way to do some native setup would be a big improvement.
